radixdlt-java/src/main/java/com/radixdlt/client/core/network/websocket/WebSocketClient.java

```diff
@@ -0,0 +1,140 @@
+package com.radixdlt.client.core.network.websocket;
+
+import com.radixdlt.client.core.network.jsonrpc.PersistentChannel;
+import io.reactivex.Observable;
+import io.reactivex.subjects.BehaviorSubject;
+import io.reactivex.subjects.PublishSubject;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import okhttp3.Response;
+import okhttp3.WebSocket;
+import okhttp3.WebSocketListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages the state for a single websocket.
+ */
+public class WebSocketClient implements PersistentChannel {
+	private static final Logger LOGGER = LoggerFactory.getLogger(WebSocketClient.class);
+	private final Object lock = new Object();
+	private final BehaviorSubject<WebSocketStatus> state = BehaviorSubject.createDefault(WebSocketStatus.DISCONNECTED);
+	private final Function<WebSocketListener, WebSocket> websocketFactory;
+	private final PublishSubject<String> messages = PublishSubject.create();
+	private WebSocket webSocket;
+
+	public WebSocketClient(Function<WebSocketListener, WebSocket> websocketFactory) {
+		this.websocketFactory = websocketFactory;
+		this.state
+			.filter(state -> state.equals(WebSocketStatus.FAILED))
+			.debounce(1, TimeUnit.MINUTES)
+			.subscribe(i -> {
+				synchronized (lock) {
+					if (this.state.getValue().equals(WebSocketStatus.FAILED)) {
+						this.state.onNext(WebSocketStatus.DISCONNECTED);
+					}
+				}
+			});
+	}
+
+	private WebSocket connectWebSocket() {
+		synchronized (lock) {
+			return this.websocketFactory.apply(new WebSocketListener() {
+				@Override
+				public void onOpen(WebSocket webSocket, Response response) {
+					synchronized (lock) {
+						WebSocketClient.this.state.onNext(WebSocketStatus.CONNECTED);
+					}
+				}
+
+				@Override
+				public void onMessage(WebSocket webSocket, String message) {
+					synchronized (lock) {
+						messages.onNext(message);
+					}
+				}
+
+				@Override
+				public void onClosing(WebSocket webSocket, int code, String reason) {
+					webSocket.close(1000, null);
```

Comment:
Should state not change here, given that we have just closed the socket?

Reply:
In practice I haven't seen `onClosing` being called but the state change will occur in the `onClose` callback.